node/core/pvf-checker/src/lib.rs

```diff
@@ -294,8 +304,50 @@ async fn handle_leaves_update(
 			.on_leaves_update(Some((activated.hash, pending_pvfs)), &update.deactivated);
 		metrics.on_pvf_observed(outcome.newcomers.len());
 		metrics.on_pvf_left(outcome.left_num);
-		for newcomer in outcome.newcomers {
-			initiate_precheck(state, sender, recent_block_hash, newcomer, metrics).await;
+		if outcome.newcomers.len() > 0 {
+			let session_index = match (&new_session_index, &state.latest_session) {
+				(Some((nsi, _)), _) => *nsi,
+				(None, Some(lsi)) => *lsi,
+				(None, None) => {
+					// examine_activation() always returns new session index if latest session
+					// index is not known; thus, getting `None` here means an error to obtain a
+					// session index for an activated leaf. At the point we do not know if it's a
+					// permanent error or not, so just drop it on the floor and run.
+					gum::warn!(
+						target: LOG_TARGET,
+						relay_parent = ?activated.hash,
+						"Failed to obtain session index",
+					);
+					return
+				},
+			};
+			let ext_session_info = match runtime
+				.get_session_info_by_index(sender, activated.hash, session_index)
+				.await
+			{
+				Ok(esi) => esi,
+				Err(error) => {
+					gum::warn!(
+						target: LOG_TARGET,
+						relay_parent = ?activated.hash,
+						?error,
+						"Failed to obtain session info",
+					);
+					return
+				},
+			};
+			let executor_params = &ext_session_info.session_info.executor_params;
+			for newcomer in outcome.newcomers {
+				initiate_precheck(
+					state,
+					sender,
+					recent_block_hash,
```

Comment:
as discussed this should be `activated.hash`:
```suggestion
					activated.hash,
```